### PR TITLE
Smaato: Rework multi imp support and add adpod support

### PR DESCRIPTION
**Pull request by el-chuck**
* Smaatos Ad Server doesn't fully supports multiple impressions per request, creating an individual request per valid bid request is an alternative solution. 
Related change on PB.js: prebid/Prebid.js#6814
* Adding adpod support from video endpoint
Related change PB.js: prebid/Prebid.js#7101
* Related docs PR prebid/prebid.github.io#3062

### CODE REVIEW

File: adapters/smaato/clock.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+type realClock struct{}
+type mockClock time.Time
+
+func (_ realClock) Now() time.Time {
+	return time.Now()
+}
+
+func (mockClock mockClock) Now() time.Time {
+	return time.Time(mockClock)
+}
```
Comment (SyntaxNode) [inline adapters/smaato/clock.go:18]: Please consider using the `Time` interface in `github.com/prebid/prebid-server/util/timeutil`.

File: adapters/smaato/smaatotest/params/race/video.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  "publisherId": "1100042525",
+  "adspaceId": "130563103"
```
Comment (mansinahar) [inline adapters/smaato/smaatotest/params/race/video.json:3]: Should this be `adbreakId` for video requests?

File: adapters/smaato/smaato.go [hunk 3]
```diff
@@ -49,240 +52,381 @@ type siteExtData struct {
...
-			errs = append(errs, userExtErr)
-			return nil, errs
+func (adapter *SmaatoAdapter) makeIndividualRequests(request *openrtb2.BidRequest) ([]*adapters.RequestData, []error) {
+	imps := request.Imp
+
+	requests := make([]*adapters.RequestData, 0, len(imps))
+	errors := make([]error, 0, len(imps))
+
+	for impIndex := range imps {
+		request.Imp = imps[impIndex : impIndex+1]
```
[the remark this answers is not in the document]
Comment (el-chuck) [reply adapters/smaato/smaato.go:166]: we need a slice with single imp here, not the imp

File: adapters/smaato/smaatotest/params/race/video.json (hunk 2 above)
Comment (el-chuck) [reply adapters/smaato/smaatotest/params/race/video.json:3]: no, a video imp via auction endpoint is fine and doesn't use adbreakId

**Comment by mansinahar** [comment]
@el-chuck Thanks for addressing the previous comments. I came across a few more things while going through this PR again. I think other than the above comments, the PR looks to be in a good shape

File: adapters/smaato/smaato_test.go [hunk 4]
```diff
@@ -16,5 +21,90 @@ func TestJsonSamples(t *testing.T) {
...
 
+	smaatoAdapter, _ := bidder.(*SmaatoAdapter)
+	assert.NotNil(t, smaatoAdapter.clock)
+	smaatoAdapter.clock = &mockTime{time: time.Date(2021, 6, 25, 10, 00, 0, 0, time.UTC)}
+
 	adapterstest.RunJSONBidderTest(t, "smaatotest", bidder)
 }
+
+func TestVideoWithCategoryAndDuration(t *testing.T) {
+	bidder := &SmaatoAdapter{}
```
[the remark this answers is not in the document]
Comment (el-chuck) [reply adapters/smaato/smaato_test.go:32]: The JSON tests only compare what is given in BidderResponse.Bids[].Bid from MakeBids()
We want to check BidderResponse.Bids[].BidVideo here

File: adapters/smaato/smaato.go [hunk 5]
```diff
@@ -49,240 +52,374 @@ type siteExtData struct {
...
+	var err error
+
+	if request.User != nil && request.User.Ext != nil {
+		var userExt userExt
+		var userExtRaw map[string]json.RawMessage
+
+		if err = json.Unmarshal(request.User.Ext, &userExtRaw); err != nil {
+			return &errortypes.BadInput{Message: "Invalid user.ext."}
+		}
+		if err = json.Unmarshal(request.User.Ext, &userExt); err != nil {
```
[the remark this answers is not in the document]
Comment (el-chuck) [reply adapters/smaato/smaato.go:329]: no i meant to unmarshal request.User.Ext, but we only take Data

[the remark this answers is not in the document]
Comment (mansinahar) [reply adapters/smaato/smaato.go:329]: I don't understand. Why are you trying to umarshal `request.User.Ext` twice, first to an `userExt` struct and then to a `map[string]json.RawMessage`?

[the remark this answers is not in the document]
Comment (el-chuck) [reply adapters/smaato/smaato.go:329]: There are some keys we want to copy from the map in request.User.Ext to request.User. Those are explicitly defined via type userExt and we unmarshal request.User.Ext into it to make them  accessible.
Then the other keys in request.User.Ext we want to keep, except for "data".  So we unmarshal it to the map, delete the data key, and assign it back to request.User.Ext to have a copy.

[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply adapters/smaato/smaato.go:329]: @mansinahar good point!  
@el-chuck if so maybe it makes sense to `json.Unmarshal(request.User.Ext, &userExtRaw)`, check if "data" exists and then unmarshal data only to `userExt.Data`? You can easily access all data fields and then remove it from map.

File: adapters/smaato/smaato.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply adapters/smaato/smaato.go:167]: @el-chuck would this work?
```
for _, imp := range imps {
       request.Imp = []openrtb2.Imp{imp}
       ...
}
```

File: adapters/smaato/smaato.go [hunk 6]
```diff
@@ -298,22 +442,94 @@ func parseImpressionObject(imp *openrtb2.Imp) error {
...
+
+func groupImpressionsByPod(imps []openrtb2.Imp) (map[string]([]openrtb2.Imp), []string) {
+	pods := make(map[string][]openrtb2.Imp)
+	orderKeys := make([]string, 0)
+
+	for _, imp := range imps {
+		pod := strings.Split(imp.ID, "_")[0]
+		_, present := pods[pod]
+		pods[pod] = append(pods[pod], imp)
+		if !present {
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply adapters/smaato/smaato.go]: @mansinahar @SyntaxNode guys, this function is very similar to `appnexus.go` -> `func groupByPods`. Should we move it to common functionality?

File: adapters/smaato/smaato.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (el-chuck) [reply adapters/smaato/smaato.go:167]: changed

File: adapters/smaato/smaato.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (el-chuck) [reply adapters/smaato/smaato.go]: done as proposed

**Review by mansinahar — DISMISSED** [review]

File: adapters/smaato/smaato.go [hunk 7]
```diff
@@ -26,21 +29,18 @@ const (
 
 // SmaatoAdapter describes a Smaato prebid server adapter.
 type SmaatoAdapter struct {
```
Comment (VeronikaSolovei9) [inline adapters/smaato/smaato.go:31]: Please consider to rename to `adapter`

**Review by VeronikaSolovei9 — COMMENTED** [review]
Just one minor comment, overall LGTM

File: adapters/smaato/smaato.go (hunk 7 above)
Comment (el-chuck) [reply adapters/smaato/smaato.go]: is renamed :smile:

File: adapters/smaato/smaato.go [hunk 8]
```diff
@@ -49,240 +49,374 @@ type siteExtData struct {
...
+	ttl := int64(300)
 
-			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
-				Bid:     &bid,
-				BidType: bidType,
-			})
+	if expiresAtMillis, err := strconv.ParseInt(response.Headers.Get("X-Smt-Expires"), 10, 64); err == nil {
+		nowMillis := adapter.clock.Now().UnixNano() / 1000000
+		ttl = (expiresAtMillis - nowMillis) / 1000
+		if ttl < 0 {
```
Comment (mansinahar) [inline adapters/smaato/smaato.go:245]: Can you please add a JSON test for this condition?

File: adapters/smaato/smaato.go [hunk 9]
```diff
@@ -298,22 +432,85 @@ func parseImpressionObject(imp *openrtb2.Imp) error {
...
-	return fmt.Errorf("invalid MediaType. SMAATO only supports Banner and Video. Ignoring ImpID=%s", imp.ID)
+	return &errortypes.BadInput{Message: "Invalid MediaType. Smaato only supports Banner and Video."}
 }
 
-func extractUserExtAttributes(userExt userExt, userCopy *openrtb2.User) {
-	gender := userExt.Data.Gender
-	if gender != "" {
-		userCopy.Gender = gender
+func setImpForAdBreak(imps []openrtb2.Imp) error {
+	adBreakID, err := jsonparser.GetString(imps[0].Ext, "bidder", "adbreakId")
```
Comment (mansinahar) [inline adapters/smaato/smaato.go:439]: Please consider adding a check to make sure the length of imps is greater than 0 before accessing the 0th imp

File: adapters/smaato/smaato.go [hunk 10]
```diff
@@ -298,22 +432,85 @@ func parseImpressionObject(imp *openrtb2.Imp) error {
...
 	}
 
-	keywords := userExt.Data.Keywords
-	if keywords != "" {
-		userCopy.Keywords = keywords
+	return nil
+}
+
+func setBannerDimension(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
+	if banner.W != nil && banner.H != nil {
```
Comment (mansinahar) [inline adapters/smaato/smaato.go:460]: Can you please add a JSON test for this condition as well?

**Review by mansinahar — COMMENTED** [review]
I think this PR is in a really good shape. Thank you @el-chuck for quickly addressing all our comments. There's just a couple tests that I think we can add to improve coverage

File: adapters/smaato/smaato.go (hunk 8 above)
Comment (el-chuck) [reply adapters/smaato/smaato.go:245]: sure, added

File: adapters/smaato/smaato.go (hunk 9 above)
Comment (el-chuck) [reply adapters/smaato/smaato.go:447]: it shouldn't be possible that imps is empty, but nevertheless i added a check

File: adapters/smaato/smaato.go (hunk 10 above)
Comment (el-chuck) [reply adapters/smaato/smaato.go:468]: also created a test for this

**Review by mansinahar — APPROVED** [review]

**Review by VeronikaSolovei9 — APPROVED** [review]

**Pull request merged**